File: geekbang/algorithm-models/2.1-前缀和、数组计数代码模板/numberOfSubarrays.js
// JavaScript
// LeetCode 1248 统计优美子数组
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var numberOfSubarrays = function (nums, k) {
  var n = nums.length;
  nums.unshift(0);
  var s = new Array(n + 1).fill(0); // 下标0~n
  for (let i = 1; i <= n; i++) s[i] = s[i - 1] + (nums[i] % 2);

  var count = new Array(n + 1).fill(0); // 下标0~n
  for (let i = 0; i <= n; i++) count[s[i]] += 1;

  // s[i] + s[j] = k
  // Home many s[j] equals s[i] - k?
  var ans = 0;
  for (let i = 1; i <= n; i++) if (s[i] - k >= 0) ans += count[s[i] - k];
  return ans;
};

File: geekbang/algorithm-models/2.1-前缀和、数组计数代码模板/numberOfSubarrays.cpp
// C/C++
// LeetCode 1248 统计优美子数组
class Solution {
public:
    int numberOfSubarrays(vector<int>& nums, int k) {
        // 开头插入一个0，转化成下标1~n
        int n = nums.size();
        nums.insert(nums.begin(), 0);
        // 前缀和，下标0~n
        vector<int> sum(n + 1, 0);
        for (int i = 1; i <= n; i++)
            sum[i] = sum[i - 1] + nums[i] % 2;
        // 计数，下标0~n
        vector<int> count(n + 1, 0);
        for (int i = 0; i <= n; i++)
            count[sum[i]]++;
        int ans = 0;
        for (int i = 0; i < nums.size(); i++)
            ans += sum[i] >= k ? count[sum[i] - k] : 0;
        return ans;
    }
};
